components/camel-univocity-parsers/src/main/java/org/apache/camel/dataformat/univocity/Marshaller.java

```diff
@@ -86,7 +88,9 @@ private void writeRow(Exchange exchange, Object row, W writer) throws NoTypeConv
         Map<?, ?> map = convertToMandatoryType(exchange, Map.class, row);
         if (adaptHeaders) {
             for (Object key : map.keySet()) {
-                headers.add(convertToMandatoryType(exchange, String.class, key));

```

Comment:
Possibly the test case i mentioned in the issue will fail, because it tried pretty much similar what you mentioned.
will try again.